### ISSUE TITLE
correct pre- and post processing

### ISSUE THREAD

**Issue by Laubeee** [daniegr/EfficientPose#25]
Hi
I'm interested in using your model (tried EfficientPoseII), however I get quite strange outputs such that the detections are nowhere near the real keypoints... am I missing some pre- or post-processing step?

I am resizing and 0-padding to 368x368
and simply taking the max value out of the 368x368 output layers..

```python
import cv2
import numpy as np
import math
from matplotlib import pyplot as plt
%matplotlib inline

from tensorflow.python.platform.gfile import FastGFile
from tensorflow.compat.v1 import GraphDef
from tensorflow.compat.v1.keras.backend import get_session
from tensorflow import import_graph_def

# load model
f = FastGFile("../models/efficient_pose/models/tensorflow/EfficientPoseII.pb", 'rb')
graph_def = GraphDef()
graph_def.ParseFromString(f.read())
f.close()
model = get_session()
model.graph.as_default()
import_graph_def(graph_def)

# take a photo
cap = cv2.VideoCapture(0)
_, frame = cap.read()
cap.release()

#swap bgr2rgb
frame = np.flip(frame, 2)

# resize
f = 368 / max(frame.shape)
frame = cv2.resize(frame, (0, 0), fx=f, fy=f)

# pad
inblob = np.zeros((368,368,3), dtype="int")
d1 = int((368-frame.shape[0])/2)
d2 = int((368-frame.shape[1])/2)
inblob[d1:368-d1,d2:368-d2] = frame

# infer
output_tensor = model.graph.get_tensor_by_name('import/upscaled_confs/BiasAdd:0')
batch_outputs = model.run(output_tensor, {'import/input_res1:0': [inblob]})

# extract keypoints, draw into 0-padded image and display
img = inblob.copy()
body_parts = ['head_top', 'upper_neck', 'right_shoulder', 'right_elbow', 'right_wrist', 'thorax', 'left_shoulder', 'left_elbow', 'left_wrist', 'pelvis', 'right_hip', 'right_knee', 'right_ankle', 'left_hip', 'left_knee', 'left_ankle']
for i in range(batch_outputs.shape[3]):
    loc = np.argmax(batch_outputs[0,:,:,i])
    py = int(math.floor(loc / 368))
    px = loc % 368
    score = batch_outputs[0,py,px,i]
    print(i, body_parts[i], px, py, score)
    img = cv2.circle(img, (px, py), 2, (255,0,0), thickness=2)
    img = cv2.putText(img, str(i), (px, py), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0,0,255), 1, lineType=cv2.LINE_AA)
plt.figure(figsize=(10,10))
plt.imshow(img)
```

**Comment by daniegr** [comment]
@Laubeee: I am glad to hear about your interest in utilizing the EfficientPose framework. Regarding your implementation, I suppose you are missing the step normalizing pixel values according to the expectations of the EfficientNet backbones. Please have a look at lines 251-255 in utils/helpers.py.

**Comment by Laubeee** [comment]
oh, I see, I missed that. Thanks!
Do you know what it does exactly? the tf-docs aren't helpful... looks like they removed the functionality (so in newer versions it doesn't manipulate anything) and on git someone states such preprocessing should be part of the model in the future..

**Comment by daniegr** [comment]
The source code hopefully provides you with some insight: https://www.github.com/keras-team/keras-applications/tree/master/keras_applications%2Fimagenet_utils.py. The preprocessing utilizes statistics (mean and std) of the ImageNet dataset to zero-center the pixel values.
